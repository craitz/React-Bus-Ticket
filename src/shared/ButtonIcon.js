import React from 'react';
import { Button } from 'react-bootstrap';
import FontAwesome from 'react-fontawesome';

export const ButtonIcon = ({ label, icon, onClick, className, type }) =>
  <Button type={type} className={className} onClick={onClick}>
    <FontAwesome name={icon} />
    {(label.length > 0) && <span className="text-after-icon">{label}</span>}
  </Button>

export const ButtonIconFit = ({ labelAll, labelXs, icon, onClick, className, type }) =>
  <Button type={type} className={className} onClick={onClick}>
    <FontAwesome name={icon} />
    <span className="text-after-icon hidden-xs">{labelAll}</span>
    <span className="text-after-icon hidden-sm hidden-md hidden-lg">{labelXs}</span>
  </Button>
