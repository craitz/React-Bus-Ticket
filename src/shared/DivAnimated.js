import React from 'react';

const DivAnimated = ({ children, className }) =>
  <div className={`${className} animated bounceInLeft`}>
    {children}
  </div>

export default DivAnimated;