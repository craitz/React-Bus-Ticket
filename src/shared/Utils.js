export const ValidationStatus = {
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error',
  NONE: null
};

export const SavingStatus = {
  SAVING: 'SAVING',
  FEEDBACK: 'FEEDBACK',
  DONE: 'DONE'
};

export const PesquisaPassagensField = {
  NOME: 'nome',
  COMPRA: 'compra',
  LINHA: 'linha',
  SAIDA: 'saida'
}

export const LoginFields = {
  EMAIL: 'email',
  SENHA: 'senha'
}

export const PoltronaStatus = {
  RESERVED: 'reserved',
  SELECTED: 'selected',
  FREE: 'free'
}

export const deepCopy = (array) => {
  return JSON.parse(JSON.stringify(array));
}

export const emailToFirebaseKey = (email) => {
  return email.replace('.', '!%^&*');
};

export const firebaseKeyToEmail = (key) => {
  return key.replace("!%^&*", '.');
};

export const SequenceArray = size => [...Array(size).keys()].map(i => ++i);
export const DateNowBr = new Date().toLocaleDateString('pt-BR');
export const DateBr = date => date.toLocaleDateString('pt-BR');
export const dateToFirebase = (text) => text.replace(/\//gi, '');
export const timeToFirebase = (text) => text.replace(/:/gi, '');

export const buildIsoDate = (data, hora) => {
  const day = data.slice(0, 2);
  const month = data.slice(3, 5);
  const year = data.slice(6);
  const hour = hora.slice(0, 2);
  const minute = hora.slice(3, 5);

  return `${year}-${month}-${day}T${hour}:${minute}:00.000Z`;
};

export const objToArray = (obj) => {
  return Object.keys(obj).map((key) => obj[key]);
};

export const buildDateObj = (data, hora) => {
  const iso = buildIsoDate(data, hora);
  return Date.parse(iso);
}

export const validateRequired = (field) => {
  if (field.isPristine) {
    return null;
  } else if (field.text.length > 0) {
    return {
      state: ValidationStatus.SUCCESS,
      text: ''
    }
  } else {
    return {
      state: ValidationStatus.ERROR,
      text: 'Campo obrigat√≥rio!'
    }
  }
}
