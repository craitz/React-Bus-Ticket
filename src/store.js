import { applyMiddleware, createStore } from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';
import promise from 'redux-promise-middleware';
import reducer from './reducers';

const configureStore = () => {
  const logger = createLogger({
    collapsed: true,
    duration: true,
    diff: true,
    level: ''
  });

  const middleware = applyMiddleware(promise(), thunk, logger);

  return createStore(reducer, middleware);
}

export default configureStore();
