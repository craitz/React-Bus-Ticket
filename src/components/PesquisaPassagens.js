import React, { Component } from 'react';
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux';
import { withAuth } from '../shared/hoc';
import {
  Table,
  NavItem,
  Glyphicon,
  Label,
  Col,
  FormControl,
  Pagination,
  Jumbotron
} from 'react-bootstrap';
import TooltipOverlay from '../shared/TooltipOverlay';
import * as actions from '../actions/pesquisaPassagens.actions';
import { firebaseHelper } from '../shared/FirebaseHelper';
import * as utils from '../shared/Utils';
import DivAnimated from '../shared/DivAnimated'
import FontAwesome from 'react-fontawesome';
import { PageHeader, PageHeaderItem } from '../shared/PageHeader';

export const totalPageItems = 12;

const TableHeader = ({ className, label, onClick, sortIcon }) =>
  <th className={className}>
    <span onClick={onClick}>{label}</span>
    {sortIcon}
  </th>

const TableColFilter = ({ tooltip, value, onChange }) =>
  <td>
    <TooltipOverlay text={tooltip} position="top">
      <FormControl value={value} onChange={onChange} />
    </TooltipOverlay>
  </td>

export const getGlyphEx = (fieldName, sort) => {
  if ((fieldName.length > 0) && (sort && (sort.field === fieldName))) {
    return (
      <Glyphicon
        glyph={(sort.direction) ? "sort-by-attributes" : "sort-by-attributes-alt"}
        className="th-icon"
      />
    );
  } else {
    return null;
  }
};

export const sortByFieldEx = (passagensBackup, sort) => {
  if (!passagensBackup) {
    return [];
  }

  const passagensOrdenadas = utils.deepCopy(passagensBackup);

  if (!sort || (sort.field.length === 0)) {
    return passagensOrdenadas;
  }

  passagensOrdenadas.sort((a, b) => {
    let objA = '';
    let objB = '';

    switch (sort.field) {
      case utils.PesquisaPassagensField.LINHA: {
        objA = a.origem.concat(a.destino);
        objB = b.origem.concat(b.destino);
        break;
      }
      case utils.PesquisaPassagensField.SAIDA: {
        objA = utils.buildIsoDate(a.data, a.horario);
        objB = utils.buildIsoDate(b.data, b.horario);
        break;
      }
      case utils.PesquisaPassagensField.COMPRA: {
        objA = a.dataCompra;
        objB = b.dataCompra;
        break;
      }
      default: {
      }
    }

    if (sort.direction) {
      if (objA > objB) {
        return 1;
      }
      if (objA < objB) {
        return -1;
      }
      return 0;
    } else {
      if (objA < objB) {
        return 1;
      }
      if (objA > objB) {
        return -1;
      }
      return 0;
    }
  });

  return passagensOrdenadas;
};

export const filtraPassagensEx = (passagensComOrdenacao, filter) => {
  if (!passagensComOrdenacao || !filter) {
    return [];
  }

  const passagensFiltradas = passagensComOrdenacao.filter((passagem) => {
    const compra = passagem.dataCompra;
    const saida = passagem.data.concat(passagem.horario);
    const linha = passagem.origem.toLowerCase().concat(passagem.destino.toLowerCase());
    const poltronas = passagem.poltrona.split(' - ');
    let include = true;

    if (filter.compra.length > 0) {
      include = include && (compra.includes(filter.compra));
    }

    if (filter.linha.length > 0) {
      include = include && (linha.includes(filter.linha));
    }

    if (filter.saida.length > 0) {
      include = include && (saida.includes(filter.saida));
    }

    if (filter.poltrona.length > 0) {
      const poltronasFiltro = filter.poltrona.split(',');

      include = include &&
        poltronas.find((poltrona) => {
          return poltronasFiltro.includes(poltrona);
        });
    }

    return include;
  });

  return passagensFiltradas;
};

export class PesquisaPassagens extends Component {
  constructor(props) {
    super(props);
    this.handleClickDataCompra = this.handleClickDataCompra.bind(this);
    this.handleClickLinha = this.handleClickLinha.bind(this);
    this.handleClickSaida = this.handleClickSaida.bind(this);
    this.handleChangeFilterDataCompra = this.handleChangeFilterDataCompra.bind(this);
    this.handleChangeFilterLinha = this.handleChangeFilterLinha.bind(this);
    this.handleChangeFilterSaida = this.handleChangeFilterSaida.bind(this);
    this.handleChangeFilterPoltrona = this.handleChangeFilterPoltrona.bind(this);
    this.handleReset = this.handleReset.bind(this);
    this.handleComprarPassagem = this.handleComprarPassagem.bind(this);
    this.handleSelectPage = this.handleSelectPage.bind(this);
    this.passagensBackup = [];
  }

  getGlyph(fieldName) {
    return getGlyphEx(fieldName, this.props.consulta.sort);
  }

  sortByField() {
    return sortByFieldEx(this.passagensBackup, this.props.consulta.sort)
  }

  filtraPassagens(passagensComOrdenacao) {
    return filtraPassagensEx(passagensComOrdenacao, this.props.consulta.filter);
  }

  getTotalPages(passagens) {
    if (!passagens || (passagens.length === 0)) {
      return 0;
    }

    const total = passagens.length;
    const quot = parseInt((total / totalPageItems), 10);
    const rest = total % totalPageItems
    return (rest > 0) ? (quot + 1) : quot;
  }

  buildPage(passagemOrdenadaFiltrada) {
    const { dispatch, consulta } = this.props;
    const { activePage } = consulta;
    let newActivePage = activePage;

    if (!passagemOrdenadaFiltrada || (passagemOrdenadaFiltrada.length === 0)) {
      dispatch(actions.setActivePage(0));
      dispatch(actions.setPage([]));
      return;
    }

    const total = this.getTotalPages(passagemOrdenadaFiltrada);

    if (activePage === 0) {
      newActivePage = 1;
    } else if (activePage > total) {
      newActivePage = total;
    }

    const start = (newActivePage - 1) * totalPageItems;
    const end = newActivePage * totalPageItems;
    const newPage = utils.deepCopy(passagemOrdenadaFiltrada.slice(start, end));

    dispatch(actions.setActivePage(newActivePage));
    dispatch(actions.setPage(newPage));
  }

  componentWillMount() {
    const emailFirebase = utils.emailToFirebaseKey(firebaseHelper.getUserEmail());
    const ref = `passagens/${emailFirebase}`;

    firebaseHelper.fetch(ref).then((passagensObj) => {
      const passagensArray = utils.objToArray(passagensObj);
      this.passagensBackup = utils.deepCopy(passagensArray);
      this.handleReset();
    });
  }

  // shouldComponentUpdate(nextProps, nextState) {
  //   return true;
  // }

  handleSelectPage(eventKey) {
    const { passagens, dispatch } = this.props;

    dispatch(actions.setActivePageThunk(eventKey))
      .then(() => {
        this.buildPage(passagens);
      });
  }

  handleReset() {
    this.props.dispatch(actions.resetConsultaThunk())
      .then(() => {
        this.aplicaPesqusia();
      });
  }

  handleComprarPassagem(event) {
    event.preventDefault();
    this.props.history.push('/comprar');
  }

  aplicaPesqusia() {
    const arrSorted = this.sortByField();
    const arrSortedFiltered = this.filtraPassagens(arrSorted);
    this.buildPage(arrSortedFiltered);
    this.props.dispatch(actions.setPassagens(arrSortedFiltered));
  }

  handleClickDataCompra(event) {
    const { consulta, dispatch } = this.props;
    const { sort } = consulta;

    if (sort.field === utils.PesquisaPassagensField.COMPRA) {
      dispatch(actions.setSortDirectionThunk(!sort.direction))
        .then(() => {
          this.aplicaPesqusia();
        });
    } else {
      dispatch(actions.setSortThunk(utils.PesquisaPassagensField.COMPRA, true))
        .then(() => {
          this.aplicaPesqusia();
        });
    }
  }

  handleClickLinha(event) {
    const { consulta, dispatch } = this.props;
    const { sort } = consulta;

    if (sort.field === utils.PesquisaPassagensField.LINHA) {
      dispatch(actions.setSortDirectionThunk(!sort.direction))
        .then(() => {
          this.aplicaPesqusia();
        });
    } else {
      dispatch(actions.setSortThunk(utils.PesquisaPassagensField.LINHA, true))
        .then(() => {
          this.aplicaPesqusia();
        });
    }
  }

  handleClickSaida(event) {
    const { consulta, dispatch } = this.props;
    const { sort } = consulta;

    if (sort.field === utils.PesquisaPassagensField.SAIDA) {
      dispatch(actions.setSortDirectionThunk(!sort.direction))
        .then(() => {
          this.aplicaPesqusia();
        });
    } else {
      dispatch(actions.setSortThunk(utils.PesquisaPassagensField.SAIDA, true))
        .then(() => {
          this.aplicaPesqusia();
        });
    }
  }

  handleChangeFilterDataCompra(event) {
    const value = event.target.value;
    this.props.dispatch(actions.setFilterCompraThunk(value))
      .then(() => {
        this.aplicaPesqusia();
      });
  }

  handleChangeFilterLinha(event) {
    const value = event.target.value.toLowerCase();
    this.props.dispatch(actions.setFilterLinhaThunk(value))
      .then(() => {
        this.aplicaPesqusia();
      });
  }

  handleChangeFilterSaida(event) {
    const value = event.target.value;
    this.props.dispatch(actions.setFilterSaidaThunk(value))
      .then(() => {
        this.aplicaPesqusia();
      });
  }

  handleChangeFilterPoltrona(event) {
    const value = event.target.value;
    this.props.dispatch(actions.setFilterPoltronaThunk(value))
      .then(() => {
        this.aplicaPesqusia();
      });
  }

  render() {
    const { passagens, consulta } = this.props;
    const { filter, activePage, page } = consulta;
    const fields = utils.PesquisaPassagensField;
    const totalPages = this.getTotalPages(passagens);

    return (
      <div className="pesquisar-passagens-container">
        <PageHeader title="Histórico de compras" className="hidden-xs" />
        <PageHeader title="Histórico" className="visible-xs" />
        {/*<NavItem className="resultados hidden-sm" href="#">
            <FontAwesome className={passagens.length > 0 ? "icon text-success" : "icon text-danger"} name={passagens.length > 0 ? "check" : "times"} />
            <span className="text-after-icon">{passagens.length} resultados encontrados</span>
          </NavItem>
          <PageHeaderItem tooltip="Comprar passagens" glyph="shopping-cart" onClick={this.handleComprarPassagem} />
          <PageHeaderItem tooltip="Limpar filtros" glyph="eraser" onClick={this.handleReset} />
        </PageHeader>*/}
        <DivAnimated>
          <Col md={10} mdOffset={1}>
            <Pagination
              bsSize="medium"
              items={totalPages > 1 ? totalPages : 0}
              activePage={activePage}
              onSelect={this.handleSelectPage}
              className="pagination-pesquisa mui--z2"
            />
            <Jumbotron className="mui--z2">
              <Table responsive hover>
                <thead>
                  <tr>
                    <TableHeader
                      className="th-data-compra"
                      onClick={this.handleClickDataCompra}
                      label="Data da compra"
                      sortIcon={this.getGlyph(fields.COMPRA)}
                    />
                    <TableHeader
                      className="th-linha"
                      onClick={this.handleClickLinha}
                      label="Linha"
                      sortIcon={this.getGlyph(fields.LINHA)}
                    />
                    <TableHeader
                      className="th-saida"
                      onClick={this.handleClickSaida}
                      label="Saída"
                      sortIcon={this.getGlyph(fields.SAIDA)}
                    />
                    <th>Poltrona(s)</th>
                  </tr>
                </thead>
                <tbody className="text-left">
                  <tr>
                    <TableColFilter
                      tooltip="Pesquisar pela data da compra da passagem"
                      value={filter.compra}
                      onChange={this.handleChangeFilterDataCompra}
                    />
                    <TableColFilter
                      tooltip="Pesquisar pela origem e destino da viagem"
                      value={filter.linha}
                      onChange={this.handleChangeFilterLinha}
                    />
                    <TableColFilter
                      tooltip="Pesquisar pela data e horário de saída"
                      value={filter.saida}
                      onChange={this.handleChangeFilterSaida}
                    />
                    <TableColFilter
                      tooltip="Pesquisar por poltrona(s). Se necessário, separe por vírgulas"
                      value={filter.poltrona}
                      onChange={this.handleChangeFilterPoltrona}
                    />
                  </tr>
                  {page.map((value, index) =>
                    <tr key={index}>
                      <td>
                        {value.dataCompra}
                      </td>
                      <td>
                        <span>{value.origem}</span>
                        <Glyphicon glyph="arrow-right" className="icon-table arrow-trajeto" />
                        <span>{value.destino}</span>
                      </td>
                      <td>
                        <Glyphicon glyph="calendar" className="icon-table data" />
                        <span>{value.data}</span>
                        <Glyphicon glyph="time" className="icon-table horario" />
                        <span>{value.horario}</span>
                      </td>
                      <td>
                        {value.poltrona.split(' - ').map((value, index) =>
                          <Label key={index} bsStyle="danger" className="label-poltrona">{value}</Label>
                        )}
                      </td>
                    </tr>
                  )}
                </tbody>
              </Table>
            </Jumbotron>
          </Col>
        </DivAnimated>

      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    passagens: state.pesquisaPassagensState.passagens,
    consulta: state.pesquisaPassagensState.consulta
  };
};

const PesquisaPassagensWithRouter = withRouter(PesquisaPassagens);
const PesquisaPassagensWithRouterAndAuth = withAuth(PesquisaPassagensWithRouter);
export default connect(mapStateToProps)(PesquisaPassagensWithRouterAndAuth);
