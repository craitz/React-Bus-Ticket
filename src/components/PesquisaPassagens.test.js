import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux';
import store from '../store';
import renderer from 'react-test-renderer';
import Immutable from 'seamless-immutable';
import { Glyphicon } from 'react-bootstrap';
import PesquisaPassagens, {
  Consulta,
  getGlyph,
  sortByField,
  filtraPassagens,
  resetConsulta,
  getTotalPages,
  setPage,
  totalPageItems
} from './PesquisaPassagens';
import { PesquisaPassagensField } from '../shared/Utils';
import { SequenceArray } from '../shared/Utils';

const passagem = Immutable({
  cpf: "032.445.687-22",
  data: "23/08/2017",
  dataCompra: "26/06/2017",
  destino: "Florianópolis (SC)",
  email: "guest@busticket.com",
  horario: "10:00",
  nome: "Darth Vader",
  origem: "Rio de Janeiro (RJ)",
  poltrona: "35"
});

const merge = (obj) => {
  return passagem.merge(obj, { deep: true });
}

describe('PesquisaPassagens - VIEW', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Router>
        <Provider store={store}>
          <PesquisaPassagens />
        </Provider>
      </Router>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});


describe('PesquisaPassagens - setPage()', () => {
  it('retorna [] se o array está vazio', () => {
    setPage([]);
    expect(Consulta.activePage).toBe(0);
    expect(Consulta.page).toEqual([]);
  });

  it('retorna [] se o array é nulo', () => {
    setPage(null);
    expect(Consulta.activePage).toBe(0);
    expect(Consulta.page).toEqual([]);
  });

  it('retorna [] se o array é undefined', () => {
    setPage(undefined);
    expect(Consulta.activePage).toBe(0);
    expect(Consulta.page).toEqual([]);
  });

  it('altera o activePage se era zero', () => {
    Consulta.activePage = 0;
    setPage(SequenceArray(20));
    expect(Consulta.activePage).toBe(1);
    expect(Consulta.page.length).toBe(totalPageItems);
  });

  it('altera o activePage quando alista dininui', () => {
    Consulta.activePage = 4;
    setPage(SequenceArray(20));
    expect(Consulta.activePage).toBe(2);
    expect(Consulta.page.length).toBe(8);
  });

  it('mantem o activePage quando a lista aumenta', () => {
    Consulta.activePage = 1;
    setPage(SequenceArray(121));
    expect(Consulta.activePage).toBe(1);
    expect(Consulta.page.length).toBe(totalPageItems);
  });
});

describe('PesquisaPassagens - getTotalPages()', () => {
  it('retorna [] se o array está vazio', () => {
    expect(getTotalPages([])).toBe(0);
  });

  it('retorna [] se o array é nulo', () => {
    expect(getTotalPages(null)).toBe(0);
  });

  it('retorna [] se o array é undefined', () => {
    expect(getTotalPages(undefined)).toBe(0);
  });

  it('retorna número de páginas quebrado', () => {
    expect(getTotalPages(SequenceArray(122))).toBe(11);
  });

  it('retorna número de páginas exato', () => {
    expect(getTotalPages(SequenceArray(240))).toBe(20);
  });

  it('retorna apenas 1 página', () => {
    expect(getTotalPages(SequenceArray(1))).toBe(1);
  });
});

describe('PesquisaPassagens - filtraPassagens()', () => {
  it('retorna [] se o array está vazio', () => {
    expect(filtraPassagens([])).toEqual([]);
  });

  it('retorna [] se o array é nulo', () => {
    expect(filtraPassagens(null)).toEqual([]);
  });

  it('retorna [] se o array é undefined', () => {
    expect(filtraPassagens(undefined)).toEqual([]);
  });

  it('filtra por nome', () => {
    const passagem1 = merge({
      nome: 'AABBCC',
    });
    const passagem2 = merge({
      nome: 'AABBBCC',
    });
    const passagens = [passagem1, passagem2];

    Consulta.filter.nome = 'bbb';

    expect(filtraPassagens(passagens)[0].nome).toBe('AABBBCC');
  });

  it('filtra por data da compra', () => {
    const passagem1 = merge({
      nome: 'AABBCC',
      dataCompra: '18/09/1974',
    });
    const passagem2 = merge({
      nome: 'AABBBCC',
      dataCompra: '22/09/1978',
    });
    const passagens = [passagem1, passagem2];

    resetConsulta();
    Consulta.filter.compra = '78';

    expect(filtraPassagens(passagens)[0].nome).toBe('AABBBCC');
  });

  it('filtra por linha', () => {
    const passagem1 = merge({
      nome: 'AABBCC',
      origem: 'Florianópolis (SC)',
      destino: 'Rio de Janeiro (RJ)',
    });
    const passagem2 = merge({
      nome: 'AABBBCC',
      origem: 'Fortaleza (CE)',
      destino: 'Aracajú (SE)',
    });

    const passagens = [passagem1, passagem2];

    resetConsulta();
    Consulta.filter.linha = 'ale';

    expect(filtraPassagens(passagens)[0].nome).toBe('AABBBCC');
  });

  it('filtra por saida', () => {
    const passagem1 = merge({
      nome: 'AABBCC',
      data: '30/10/2017',
      horario: '22:00'
    });
    const passagem2 = merge({
      nome: 'AABBBCC',
      data: '14/07/2017',
      horario: '06:00'
    });
    const passagens = [passagem1, passagem2];

    resetConsulta();
    Consulta.filter.saida = '07';

    expect(filtraPassagens(passagens)[0].nome).toBe('AABBBCC');
  });

  it('filtra por poltrona', () => {
    const passagem1 = passagem.merge({
      nome: 'AABBCC',
      poltrona: '1 - 41 - 42'
    });
    const passagem2 = passagem.merge({
      nome: 'AABBBCC',
      poltrona: '39 - 40 - 42'
    });
    const passagens = [passagem1, passagem2];

    resetConsulta();
    Consulta.filter.poltrona = '39,40';

    expect(filtraPassagens(passagens)[0].nome).toBe('AABBBCC');
  });
});

describe('PesquisaPassagens - sortByField()', () => {
  it('retorna [] se o array está vazio', () => {
    expect(sortByField([])).toEqual([]);
  });

  it('retorna [] se o array é nulo', () => {
    expect(sortByField(null)).toEqual([]);
  });

  it('retorna [] se o array é undefined', () => {
    expect(sortByField(undefined)).toEqual([]);
  });

  it('ordena por nome', () => {
    const passagem1 = passagem.merge({ nome: 'CBA' });
    const passagem2 = passagem.merge({ nome: 'ABC' });
    const passagens = [passagem1, passagem2];
    Consulta.sort.field = PesquisaPassagensField.NOME;
    expect(sortByField(passagens)[0].nome).toBe('ABC');
  });

  it('ordena por data da compra', () => {
    const passagem1 = passagem.merge({ dataCompra: '22/09/1978' });
    const passagem2 = passagem.merge({ dataCompra: '18/09/1974' });
    const passagens = [passagem1, passagem2];
    Consulta.sort.field = PesquisaPassagensField.COMPRA;
    expect(sortByField(passagens)[0].dataCompra).toBe('18/09/1974');
  });

  it('ordena por linha', () => {
    const passagem1 = passagem.merge({
      origem: 'Rio de Janeiro (RJ)',
      destino: 'Florianópolis (SC)'
    });
    const passagem2 = passagem.merge({
      origem: 'Florianópolis (SC)',
      destino: 'Rio de Janeiro (RJ)'
    });
    const passagens = [passagem1, passagem2];
    Consulta.sort.field = PesquisaPassagensField.LINHA;
    expect(sortByField(passagens)[0].origem).toBe('Florianópolis (SC)');
  });

  it('ordena por saída', () => {
    const passagem1 = passagem.merge({
      data: '22/09/1978',
      horario: '18:00'
    });
    const passagem2 = passagem.merge({
      data: '22/09/1978',
      horario: '06:00'
    });
    const passagens = [passagem1, passagem2];
    Consulta.sort.field = PesquisaPassagensField.SAIDA;
    expect(sortByField(passagens)[0].horario).toBe('06:00');
  });
});

describe('PesquisaPassagens - getGlyph()', () => {
  it('retorna null se field não está sendo pesquisado', () => {
    Consulta.sort.field = PesquisaPassagensField.COMPRA;
    expect(getGlyph(PesquisaPassagensField.LINHA)).toBeNull();
  });

  it('retorna null se field vazio', () => {
    expect(getGlyph('')).toBeNull();
  });

  it('retorna null se field invalido', () => {
    expect(getGlyph('INVALID_FIELD')).toBeNull();
  });

  it('retorna gliph se field é igual', () => {
    Consulta.sort.field = PesquisaPassagensField.COMPRA;
    Consulta.sort.direction = false;
    expect(getGlyph(PesquisaPassagensField.COMPRA))
      .toEqual(<Glyphicon bsClass="glyphicon" className="th-icon" glyph="sort-by-attributes-alt" />);
  });
});