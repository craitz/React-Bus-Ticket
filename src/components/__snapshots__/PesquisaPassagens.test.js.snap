// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PesquisaPassagens - VIEW renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <PageWithAuth
    consulta={
        Object {
            "activePage": 1,
            "filter": Object {
              "compra": "",
              "linha": "",
              "nome": "",
              "poltrona": "",
              "saida": "",
            },
            "page": Array [],
            "sort": Object {
              "direction": true,
              "field": "",
            },
          }
    }
    dispatch={[Function]}
    passagens={Array []}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "notify": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            },
            "unsubscribe": null,
          }
    }
/>,
  "nodes": Array [
    <PageWithAuth
      consulta={
            Object {
                  "activePage": 1,
                  "filter": Object {
                    "compra": "",
                    "linha": "",
                    "nome": "",
                    "poltrona": "",
                    "saida": "",
                  },
                  "page": Array [],
                  "sort": Object {
                    "direction": true,
                    "field": "",
                  },
                }
      }
      dispatch={[Function]}
      passagens={Array []}
      store={
            Object {
                  "dispatch": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                  Symbol(observable): [Function],
                }
      }
      storeSubscription={
            Subscription {
                  "listeners": Object {
                    "notify": [Function],
                  },
                  "onStateChange": [Function],
                  "parentSub": undefined,
                  "store": Object {
                    "dispatch": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                    Symbol(observable): [Function],
                  },
                  "unsubscribe": null,
                }
      }
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Connect(PageWithAuth)
        store={
                Object {
                        "dispatch": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                        Symbol(observable): [Function],
                      }
        }
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Connect {
        "_reactInternalInstance": [Circular],
        "context": Object {
          "store": undefined,
          "storeSubscription": undefined,
        },
        "notifyNestedSubs": [Function],
        "props": Object {
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
        },
        "propsMode": true,
        "refs": Object {},
        "renderCount": 0,
        "selector": Object {
          "error": null,
          "props": Object {
            "consulta": Object {
              "activePage": 1,
              "filter": Object {
                "compra": "",
                "linha": "",
                "nome": "",
                "poltrona": "",
                "saida": "",
              },
              "page": Array [],
              "sort": Object {
                "direction": true,
                "field": "",
              },
            },
            "dispatch": [Function],
            "passagens": Array [],
            "store": Object {
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            },
          },
          "run": [Function],
          "shouldComponentUpdate": false,
        },
        "setWrappedInstance": [Function],
        "state": Object {},
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "subscription": Subscription {
          "listeners": Object {
            "notify": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": null,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "version": 0,
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <PageWithAuth
          consulta={
                    Object {
                              "activePage": 1,
                              "filter": Object {
                                "compra": "",
                                "linha": "",
                                "nome": "",
                                "poltrona": "",
                                "saida": "",
                              },
                              "page": Array [],
                              "sort": Object {
                                "direction": true,
                                "field": "",
                              },
                            }
          }
          dispatch={[Function]}
          passagens={Array []}
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          storeSubscription={
                    Subscription {
                              "listeners": Object {
                                "notify": [Function],
                              },
                              "onStateChange": [Function],
                              "parentSub": undefined,
                              "store": Object {
                                "dispatch": [Function],
                                "getState": [Function],
                                "replaceReducer": [Function],
                                "subscribe": [Function],
                                Symbol(observable): [Function],
                              },
                              "unsubscribe": null,
                            }
          }
/>,
        "_debugID": 4,
        "_renderedOutput": <PageWithAuth
          consulta={
                    Object {
                              "activePage": 1,
                              "filter": Object {
                                "compra": "",
                                "linha": "",
                                "nome": "",
                                "poltrona": "",
                                "saida": "",
                              },
                              "page": Array [],
                              "sort": Object {
                                "direction": true,
                                "field": "",
                              },
                            }
          }
          dispatch={[Function]}
          passagens={Array []}
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          storeSubscription={
                    Subscription {
                              "listeners": Object {
                                "notify": [Function],
                              },
                              "onStateChange": [Function],
                              "parentSub": undefined,
                              "store": Object {
                                "dispatch": [Function],
                                "getState": [Function],
                                "replaceReducer": [Function],
                                "subscribe": [Function],
                                Symbol(observable): [Function],
                              },
                              "unsubscribe": null,
                            }
          }
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Connect(PageWithAuth)
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
}
`;
